craft_application/application.py

```diff
@@ -630,6 +630,16 @@ def _expand_environment(self, yaml_data: dict[str, Any], build_for: str) -> None
         :param yaml_data: The project's yaml data.
         :param build_for: The architecture to build for.
         """
+        if build_for == "all":
+            build_for_arch = util.get_host_architecture()
```

Comment:
This is okay for now, but thinking long-term, I think we should consider how to handle this in `craft-parts`. Specifically,  staged packages/snaps should only be platform-independent too. (This would require a lot more work though.)